asserts/snap_asserts.go

```diff
@@ -698,3 +698,140 @@ func InitBuiltinBaseDeclaration(headers []byte) error {
+	}
+
+	// check there's an account for each developer
+	for developerID := range snapdev.developerRanges {
+		if developerID == publisherID {
+			continue
+		}
+		_, err = db.Find(AccountType, map[string]string{"account-id": developerID})
+		if err == ErrNotFound {
+			return fmt.Errorf("snap-developer assertion for snap-id %q does not have a matching account assertion for the developer %q", snapdev.SnapID(), developerID)
+		}
+	}
+
+	return nil
+}
+
+// sanity
+var _ consistencyChecker = (*SnapDeveloper)(nil)
+
+// Prerequisites returns references to this snap-developer's prerequisite assertions.
+func (snapdev *SnapDeveloper) Prerequisites() []*Ref {
+	// Capacity for the snap-declation, the publisher and all developers.
+	refs := make([]*Ref, 0, 2+len(snapdev.developerRanges))
+
+	// snap-declaration
+	// XXX: mediate getting current series through some context object? this gets the job done for now
+	refs = append(refs, &Ref{SnapDeclarationType, []string{release.Series, snapdev.SnapID()}})
+
+	// the publisher and developers
+	publisherID := snapdev.PublisherID()
+	refs = append(refs, &Ref{AccountType, []string{publisherID}})
+	for developerID := range snapdev.developerRanges {
+		if developerID != publisherID {
+			refs = append(refs, &Ref{AccountType, []string{developerID}})
+		}
+	}
+
+	return refs
+}
+
+func assembleSnapDeveloper(assert assertionBase) (Assertion, error) {
+	developerRanges, err := checkDevelopers(assert.headers, "developers")
+	if err != nil {
+		return nil, err
+	}
+
+	return &SnapDeveloper{
+		assertionBase:   assert,
+		developerRanges: developerRanges,
+	}, nil
+}
+
+func checkDevelopers(headers map[string]interface{}, name string) (map[string][]*dateRange, error) {
+	// TODO:
+	// - reject overlapping date ranges?
+	value, ok := headers[name]
+	if !ok {
+		return nil, nil
+	}
+	developers, ok := value.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("%q must be a list of developer maps", name)
+	}
+	if len(developers) == 0 {
+		return nil, nil
+	}
+
+	developerRanges := make(map[string][]*dateRange)
+	for i, item := range developers {
+		developer, ok := item.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("%q must be a list of developer maps", name)
+		}
+		accountID, err := checkNotEmptyStringWhat(developer, "developer-id", "item")
+		if err != nil {
+			return nil, fmt.Errorf("%s[%d]'s %s", name, i, err)
+		}
+		since, err := checkRFC3339DateWhat(developer, "since", "item")
+		if err != nil {
+			return nil, fmt.Errorf("%s[%d]'s %s", name, i, err)
+		}
+		until, err := checkRFC3339DateWithDefaultWhat(developer, "until", "item", time.Time{})
+		if err != nil {
+			return nil, fmt.Errorf("%s[%d]'s %s", name, i, err)
+		}
+		if !until.IsZero() && since.After(until) {
+			return nil, fmt.Errorf(`%s[%d]'s "since" must be less than or equal to "until"`, name, i)
```

Comment:
wouldn't be better to use accountID instead of i ? (don't think we ever used the [] notations for errors here)